### ISSUE TITLE
Computed Property Returned from Model Hook Out of Sync

### ISSUE THREAD

**Issue by spinlock99** [emberjs/ember.js#13610]
I've created an example of this issue here: https://github.com/spinlock99/ember-caching-issue

Description: the example shows that a computed property returned from the route's model hook does not stay in sync with the same computed property defined on the controller.

Expected Behavior: a computed property returned from the route's model hook is expected to update based on changes made to Ember Data.

**Comment by btecu** [comment]
I think `computed` is working properly.
When using computed properties, you have to pass property dependencies if you expect it to update.

`tasks: Ember.computed(function () { this.store.peekAll('task'); })` doesn't make much sense. I'd say this is probably not the Ember way.

**Comment by xomaczar** [comment]
peekAll returns a live updating array - so anytime you push or createRecord your live array will change. 
FYI: I don't think this is the right place to ask this type of question, stackoverflow/slack is a better medium for that.

**Comment by spinlock99** [comment]
This is a bug report showing that a computed property returned from a route's model hook behaves differently from _the same_ computed property in a controller.

If Ember truly is supposed to treat a computed property _differently_ based on if it is returned from a route's model hook vs living in a controller then please close this bug report. However, this is not a question, this is an example of behavior that was very surprising to me as I expected a computed property to have the same behavior in both places.

**Comment by spinlock99** [comment]
@btecu please see https://github.com/ember-data/ember-data-live-filter-by#i-dont-want-to-use-this-addon-are-there-alternatives.

**Comment by btecu** [comment]
@spinlock99 I still don't think the issue is computed.

In your route, add `tasks: Ember.computed.alias('filteredTasks'),` and update your template to use `tasks` instead of `model`.

**Comment by spinlock99** [comment]
Thanks for taking a look at this issue @btecu. Unfortunately,  `tasks` is already defined on the route because it's in the mixin that is shared between the route and the controller. So, creating a computed alias named `tasks` on the route results in a circular reference and stack overflow:

![screen shot 2016-06-07 at 9 40 59 am](https://cloud.githubusercontent.com/assets/453431/15866564/0b2ff2c0-2c94-11e6-8189-26b54ea91619.png)

Also, I'm not trying to get this to "work". I'm trying to show that the _same_ computed property will behave differently if it is defined in a controller vs. being defined in the route and returned from the route's model hook. 

So, you have `filteredTasks` shared between the route and the controller via a mixin. I would expect this computed property to behave the same if it is passed to the template via the model hook or by directly binding to the controller. But, as you can see below, the computed property returned from the route's model hook is  _not_ bound to the underlying data where the computed property in the controller is:

![caching issue](https://cloud.githubusercontent.com/assets/453431/15866283/db03a778-2c92-11e6-9f41-6c3a67339bd3.gif)

**Comment by pixelhandler** [comment]
@spinlock99 I have not come across any documentation that recommends returning a computed property from a model hook. Typically a promise that resolves with (an array or object) the model(s) is returned from the model hook. When defering the promise resolution and returning an object with a one or more promises assigned to a property of the object you may use a computed property in your controller or component which results in re-rendering the template.

What led you to this implementation? returning a computed property as a model?

I don't this this is an Ember bug, perhaps an edge case but there are other ways to accomplish filtering items from the store for a template.

**Comment by spinlock99** [comment]
I started off with routes that looked like this (take from the (guide)[https://guides.emberjs.com/v2.6.0/routing/specifying-a-routes-model/]):

```
export default Ember.Route.extend({
  model() {
    return this.store.query('post', { favorite: true });
  }
});
```

But, this makes the user wait for the round-trip to the server every time they visit the route. So, I'm checking the store to see if there are any records to show the user and, if there's a cache hit, I'll return a filtered array so that the user sees the records that are in the store immediately and then updates when the call comes back from the server.

**Comment by pixelhandler** [comment]
@spinlock99 Perhaps ask in the slack #-help channel to so how others have solved this idea of returning what data is available immediately, rendering while fetching then re-rendering if there is more data found on the server. My inital thoughts are that you'd need to return an object from the model hook with two properties `{existing: [], fetched: []}` then in your template could use `{{#if fetched}}` and `{{#if existing}}` or some other solution. Typically I avoid these types of optimization since I trust the browser cache as well for my data. I just ask the server for the data. I still don't think this is an Ember bug.

**Comment by spinlock99** [comment]
I'd love to be able to make the assumption that the data on the server won't change unless it's mutated by the client but the concurrency model I'm dealing with is more complex than that. Both other users and automated jobs are updating the application state so etags or 304s just aren't going to cut it. 

Anyway, the point isn't that it's impossible -- or even difficult -- to figure out a workaround. The point is that I've worked with Ember at both Twitch and now at AdStage, pretty much daily for 3 years, and I was very surprised to learn that you cannot treat computed properties as first class objects. This causes maddeningly subtle bugs because the return values look fine at first and only fail when the dependencies are updated.

If this really is the Ember way then I'd be happy to submit a PR updating the documentation to make it clear that computed properties cannot be treated as data.

**Comment by Serabe** [comment]
It does not fail. It is not a bug. I already answered the question in [discuss](http://discuss.emberjs.com/t/how-can-i-bind-data-returned-from-the-model-hook-to-the-template/10730/5?u=serabe) but let me try again.

When the user hits a route in any of [the ways explained in the guides](https://guides.emberjs.com/v2.6.0/routing/), Ember follows the route lifecycle which includes three hooks directly related with models: `beforeModel`, `model` and `afterModel`. Whatever is returned from `model` will be set in the `model` property of the controller.

It is quite common to use EmberData in an Ember project and therefore is quite common to look at `model` methods that look like the following:

```js
model({ id }) {
  return this.store.findRecord('continent', id);
}
```

```js
model() {
  return this.store.findAll('continent');
}
```

We can of course have routes whose `model` method looks like:

```js
model() {
  return [
    { id: 'africa', name: 'Africa' },
    { id: 'america', name: 'America' },
    { id: 'asia', name: 'Asia'},
    { id: 'europe', name: 'Europe' },
    { id: 'oceania', name: 'Oceania' }
  ]
}
```

There is only one subtlety between returning an array and using Ember Data's query method is that `findX` returns a `Promise` and, therefore, the section [Asynchronous Routing](https://guides.emberjs.com/v2.6.0/routing/asynchronous-routing/) in the guides apply.

I know so far I have not address your question, but I am getting there. I need to talk about CPs first.

According to the [guides](https://guides.emberjs.com/v2.6.0/object-model/computed-properties/), Computed Properties

    let you declare functions as properties. You create one by defining a computed property as a function, which Ember will automatically call when you ask for the property. You can then use it the same way you would any normal, static property.

See that this function is called when you get the value (it might not be called if the value is already cached) and what you get is the returning value of the function. Therefore, given the class:

```js
import Ember from 'ember';

export default Ember.Object.extend({
  myProp: 'My String',
  
  myComputedProp: Ember.computed(function() {
    return this.get('myProp');
  })
});
```

These two values are the same: `MyObject.create().get('myProp')` and `MyObject.create().get('myComputedProp')`. The second value is not decorated in any form. [Twiddle](https://ember-twiddle.com/bd195dcf2ec60edea5acdd72276ee1ae?numColumns=2&openFiles=controllers.index.js%2Ctemplates.index.hbs)

Then, why does the template updates when we create a record of the same model or, by other means, we received one from the backend? The answer is called live arrays. Ember Data returns an array that gets updated when a new record is added. You can play with this idea [in this twiddle](https://ember-twiddle.com/3b4607944b5581a8679699cd659c2c12?numColumns=2&openFiles=twiddle.json%2Ctemplates.application.hbs).

In [this other twiddle](https://ember-twiddle.com/b05207252c6ad599c8e1a6d41b5ed33a?numColumns=2&openFiles=twiddle.json%2Ctemplates.application.hbs) we have the same code but in the `model` hook we do:

```js
model() {
  return this.store.findAll('user')
    .then((users) => users.toArray());
}
```

This gives you a plain old JavaScript array and, if you play with this last example you see that the list is not updated.

TL;DR
1. `model` hook is called only once when entering the route. It is not recalled unless you enter it again.
2. Computed Properties are a way to have `getters` and `setters` that play nice with Ember (you can look in the issues and RFCs that at some point, computed properties will be decorated plain properties.
3. Ember Data returns _live arrays_ and that is what gives the application the feeling that the `model` hook is being called several times or that return some kind of live property.

**Comment by Serabe** [comment]
I am closing this since it is not a bug in Ember. Please, if you consider that guides should explain these concepts better than they already do or you miss any explanation there, open an issue in emberjs/guides or, even better, a PR with such changes.

**Comment by spinlock99** [comment]
@serabe, thanks for taking the time to look at this issue.

Sorry to be such a pain about this but I want to make sure that we're on the same page. Let's ignore Ember Data for the moment because I do think that it's a red herring for this issue. 

My confusion came from assuming that computed properties could be treated as first class objects in Ember. But, computed properties _cannot_ be treated like normal properties because they will loose their bindings if they are returned from a function. (If you could fill in some of the thinking behind why computed properties are not referentially transparent, it would really help me with the documentation pr I'm working on.)

As an example, I've created this [twiddle](https://ember-twiddle.com/34b109a0eb6f0582a42bbddf68a3c165?openFiles=routes.application.js%2C) that just returns a simple computed property from the route's model hook:

```
import Ember from 'ember';

export default Ember.Route.extend({
  bar: 'baz',
  foo: Ember.computed('bar', function () { return this.get('bar'); }),

  model() { return this.get('foo'); },

  actions: {
    update: function () {
      this.set('bar', 'updated');
      console.log(this.get('bar'));
    }
  }
});

```

Bound to this template:

```
<h2 id="title">Welcome to Ember</h2>

Model: {{model}}<br><br>

<a {{action 'update'}}>Update</a>

```

The expected result of pressing `Update` will be to log `updated` to the console and _not_ update the template. (It continues to show `Model: baz`.) The reason for this is that computed properties will evaluate to different values depending on the context that they are executed in.

Thanks for your help.

**Comment by xomaczar** [comment]
@spinlock99 can you explain why are assuming that setting bar to the 'updated' value should not be reflected in a template? You are invalidating computed property foo which depends on bar.

**Comment by spinlock99** [comment]
@xomaczar I _did_ assume the template would update. That's why I opened this bug. If you think the template should update, perhaps we should reopen this issue ;)

**Comment by Serabe** [comment]
Addressing the issues from easier to harder to address:

1. No, this issue as is should not be reopen since it is the expected behaviour.
2. Please, move any further conversation to Slack since this is not a bug and we are generating noise in the issue tracker. This noise make the task of triaging the issues harder.
3. Computed Properties cannot be referentially transparent because they are not functions, they are properties, values. As of the functions that define them, they are (expected to be) so referentially transparent that they are cached for the last set of values of its dependencies. For this we need to extend the definition of referential transparency to include as arguments the CP dependencies meaning that given:

```js
prop: Ember.computed('a', 'b', 'c', function() { })
```

The signature of the function should be the list of arguments `a`, `b`, and `c`.


Thank you!
